rfcs/002 - Database Access.md

```diff
@@ -0,0 +1,160 @@
+```
+
+Schemas are like `type`s, but with index, join, and dbname directives included. There are three variants for the schema, the first two matching the variants for declaring a type and the third one for just deriving a schema from a type. Similarly there should be a way to derive a type from a schema:
+
+```md
+# type {name} from schema {dbname}.{type name}
+```
+
+Queries would be used by Marsha functions by the user referencing it in the description (and then the query definition would be included with the function generation, so it would be folded into the function. So different functions don't get slightly different versions of the query, there should be a pre-pass to generate the query that would be embedded into the functions so it is a SQL template given to the LLM)
+
+#### Embedded syntax
+
+Databases would still be declared top-level, and the `# schema` type would also exist, but queries would be in a new `## query` subsection for a function, instead. This would have to be after the description section but before the function examples, so the query would be defined *after* it's "use" in the description, which might be a bit weird, but if defined before the description we would need a `## description` sub-heading to make it clear again.
+
+#### References syntax
+
+Rather than coming up with a special syntax for databases or websites to mock (the [other RFC being worked on in parallel](https://github.com/alantech/marsha/pull/159)), having a references block that has a listing of references [in the Markdown style](https://www.markdownguide.org/basic-syntax/#reference-style-links) that can be referred to by number within the function blocks may provide a similar benefit, though much more reliant on the LLM to make sense of the provided data.
+
+It would still require us to connect to the URL/database/etc, to gather information to provide to the LLM, but it would be a more modular "document generator" mechanism that could be plugin-based where parsing of the markdown in the description section of the function blocks would figure out which document(s) to generate and provide to the LLM during function generation.
+
+This reduces the amount of new syntax dramatically, but eschews any database schema management within Marsha for at least the time being.
+
+Some of the desired functionality, like a shared connection object to reduce connection pressure on the database, could be done with a Marsha function told to memoize generated connection objects and other Marsha functions told to acquire the connection from said function. This places the burden on the user to "do the right thing" but keeps language simplicity and flexibility high. Potentially the user burden can be reduced by developing a standard library for Marsha and/or a module system.
+
+```md
+# func get_users(): list of user objects
+
+This function gets all user records from [the database][1] and returns them, or fails if unable to connect to the database.
```

Comment:
seems a bit tricky what substitution would need to be done here depending on each case for the final prompt for the LLM to know what to do.

Reply:
Oh, so I was thinking it would lightly transform the `[the database][1]` into `[the database](#the-database)` where `#the-database` is whatever the title is of the reference after it is turned into a markdown document itself following [this convention](https://stackoverflow.com/questions/2822089/how-to-link-to-part-of-the-same-document-in-markdown)

The exact transform would depend on the extension and how it deals with the URL (`https` should be queried and maybe it drops part of the `<body>` into the document if it's an HTML document, while a file path that ends in `.sqlite` reads that file with the sqlite3 standard library and provides a summary of the schema).